dnf-behave-tests/features/fail-safe-cache.feature

```diff
@@ -0,0 +1,308 @@
+   When I execute dnf with args "install wget"
+   Then the exit code is 0
+    And file "/var/lib/dnf/modulefailsafe/" exists
+   When I execute "ls" with args "{context.dnf.installroot}/var/lib/dnf/modulefailsafe/"
+   Then stdout is
+        """
+        nodejs:5:x86_64.yaml
+        """
+   When I execute "rm" with args "{context.dnf.installroot}/var/lib/dnf/modulefailsafe/ -r"
+   Then file "/var/lib/dnf/modulefailsafe/" does not exist
+   When I use the repository "dnf-ci-fedora-updates"
+    And I execute dnf with args "upgrade wget"
+   Then the exit code is 0
+    And file "/var/lib/dnf/modulefailsafe/" exists
+   When I execute "ls" with args "{context.dnf.installroot}/var/lib/dnf/modulefailsafe/"
+   Then stdout is
+        """
+        nodejs:5:x86_64.yaml
+        """
+
+
+@bz1616167
+@bz1623128
+Scenario: When modular RPM is installed and the modular repo is disabled and fail-safe modulemd copy is deleted, the RPM can be upgraded to a non-modular RPM
+   When I disable the repository "dnf-ci-fedora-modular-updates"
+   When I execute dnf with args "module enable nodejs:5"
+   When I execute dnf with args "install nodejs-5*"
+   Then the exit code is 0
+    And Transaction contains
+        | Action                    | Package                                            |
+        | install                   | nodejs-1:5.3.1-1.module_2011+41787af0.x86_64       |
+  Given I disable the repository "dnf-ci-fedora-modular"
+    And I execute dnf with args "clean all"
+    And I execute "rm" with args "{context.dnf.installroot}/var/lib/dnf/modulefailsafe/ -r"
+    And I use the repository "dnf-ci-fedora-modular-updates"
+   When I execute dnf with args "upgrade nodejs"
+   Then the exit code is 0
+    And Transaction is following
+        | Action                    | Package                                            |
+        | upgrade                   | nodejs-1:8.14.0-1.module_2030+42747d41.x86_64      |
+
+
+@bz1616167
+@bz1623128
+Scenario: When modular RPM is installed and the modular repo is disabled and '/var/cache/dnf' and '/var/cache/yum' are deleted, the RPM can't be upgraded to a non-modular RPM
+   When I disable the repository "dnf-ci-fedora-modular-updates"
+   When I execute dnf with args "module enable nodejs:5"
+   When I execute dnf with args "install nodejs"
+   Then the exit code is 0
+    And Transaction contains
+        | Action                    | Package                                            |
+        | install                   | nodejs-1:5.3.1-1.module_2011+41787af0.x86_64       |
+  Given I disable the repository "dnf-ci-fedora-modular"
+    And I execute dnf with args "clean all"
+    And I execute "rm" with args "{context.dnf.installroot}/var/cache/dnf {context.dnf.installroot}/var/cache/yum -r"
+    And I use the repository "dnf-ci-fedora-modular-updates"
+   When I execute dnf with args "upgrade nodejs"
+   Then the exit code is 0
+    And Transaction is empty
```

Comment:
Is this scenario failing  without FailSafe mechanism?

Reply:
Yes, it is.